packages/fxa-content-server/app/scripts/lib/survey-filter.js

```diff
@@ -31,56 +35,189 @@ export const getConditionWithKey = (conditions, key) => {
+ *
+ * @param {function} valSource a function that takes a function (a) and a comparator and apply the
+ *                             return value of (a) as the first argument to the comparator
+ * @param {function} comparator a function that compares two values and returns a boolean
+ * @param {string} key the configuration key where the condition value can be found
+ * @returns {function} a function that takes conditions and (f) (see description above) and
+ *                     returns a boolean
+ */
+export const createConditionCheckFn = (valSource) => (comparator) => (key) => (
+  conds,
+  fetchFn
+) => {
   // This turned out a little janky.  But we want to return true early if the
   // condition is not in the configs.
   const condVal = getConditionWithKey(conds, key);
   if (condVal === NONE) {
     return true;
   }
-  return valSourceCheck(valSource)(comparator)(condVal);
+  return valSource(fetchFn)(comparator)(condVal);
 };
 
-export const createAsyncConditionCheckFn = (valSourceCheck) => (comparator) => (
+// Like the function above but creates an async function because (f) can be async.
+export const createAsyncConditionCheckFn = (valSource) => (comparator) => (
   key
-) => async (conds, valSource) => {
+) => async (conds, fetchFn) => {
   const condVal = getConditionWithKey(conds, key);
   if (condVal === NONE) {
     return true;
   }
-  return (await valSourceCheck(valSource)(comparator))(condVal);
+  return (await valSource(fetchFn)(comparator))(condVal);
+};
+
+/**
+ * Creates a function that construct a UserAgent object when the given window
+ * object has a userAgent string and caches the UserAgent instance.
+ */
+export const createFetchUaFn = (window) => {
+  let ua;
+
+  return () => {
+    if (!window || !window.navigator || !window.navigator.userAgent) {
+      return undefined;
+    }
+
+    ua = ua || new UserAgent(window.navigator.userAgent);
+
+    return ua;
+  };
+};
+
+/**
+ * Creates a function that gets the signed in account from the given user model
+ * and caches it.
+ */
+export const createFetchAccountFn = (user) => {
+  let account;
+
+  return () => {
+    if (!user || !(account = account || user.getSignedInAccount())) {
+      return undefined;
+    }
+
+    return account;
+  };
+};
+
+/**
+ * Creates a function that get a property from a a model object and caches it.
+ *
+ * @param {string} getPropFnName name of the function to call to get the property
+ * @param {function} fetchFn a function to fetch the object
+ * @returns {function} an async function that returns the property or null
+ */
+export const createAsyncGetModelPropertyFn = (getPropFnName) => (fetchFn) => {
+  let prop;
+  return async () => {
+    try {
+      const m = await fetchFn();
+
+      if (m === undefined) {
+        return undefined;
+      }
+
+      prop = prop || (await m[getPropFnName]());
+
+      return prop;
+    } catch {
+      return undefined;
```

Comment:
Think you can just return without the `undefined` in these couple spots, no?

Reply:
heh, true.  I changed them from nulls to undefineds and didn't slow down to think about that.  I'll update.